accel/tcg/log_instr.c

```diff
@@ -0,0 +1,513 @@
+
+    qemu_log_instr_start(env, cl_default_trace_format | CPU_LOG_USER_ONLY, pc);
+    /* Skip this instruction commit */
+    log->force_drop = true;
+}
+
+void helper_qemu_log_instr_user_stop(CPUArchState *env, target_ulong pc)
+{
+    qemu_log_instr_stop(env, cl_default_trace_format | CPU_LOG_USER_ONLY, pc);
+}
+
+void helper_qemu_log_instr_commit(CPUArchState *env)
+{
+    _qemu_log_instr_commit(env);
+}
+
+void helper_qemu_log_instr(CPUArchState *env, target_ulong pc)
+{
+    _qemu_log_instr_instr(env, pc);
+}
+
+void qemu_log_instr_init(CPUArchState *env)
+{
+    cpu_log_instr_info_t *log = target_cpu_get_log(env);
+
+    memset((void *)log, 0, sizeof(*log));
+    log->txt_buffer = g_string_new(NULL);
+    log->regs = g_array_new(FALSE, TRUE, sizeof(log_reginfo_t));
+    log->force_drop = true;
+}
+
+/*
+ * Callback invoked by tcg when flushing on state changed
+ * TODO(am2419): should we move the flush callback here? it seems to be
+ *  instr logging-related and only used for this.
+ */
+/*
+void flush_tcg_on_log_instr_chage(void);
+void flush_tcg_on_log_instr_chage(void) {
+    warn_report("Calling real %s\r", __func__);
+    CPUState *cpu;
+    int cpu_index = 0;
+    CPU_FOREACH(cpu) {
+        warn_report("Flushing TCG for CPU %d\r", cpu_index++);
+        tb_flush(cpu);
+    }
+}
+*/
+
+/*
+ * Check for the instruction logging flags in qemu_loglevel.
+ * If we are tracing user-mode, then we need to check the per-cpu
+ * flag as multiple CPUs may be in different modes.
+ */
+bool qemu_log_instr_enabled(CPUState *cpu)
+{
+    if (qemu_loglevel_mask(CPU_LOG_USER_ONLY)) {
+        return cpu_get_log(cpu)->user_mode_tracing;
+    }
+
+    return (qemu_loglevel_mask(INSTR_LOG_MASK));
+}
+
+/*
+ * Start user tracing in the given mode.
+ * Note:
+ * Don't turn on tracing if user-mode only is selected and we are in the kernel.
+ * Make sure that qemu_loglevel doesn't get set to zero when we
+ * suspend tracing because otherwise qemu will close the logfile.
+ *
+ * Log start events are logged regardless of -dfilter and without waiting for
+ * the commit() call.
+ *
+ * We do not emit a start event to the trace if the tracing is enabled from
+ * command line or from the qemu monitor. This is because there is no associated
+ * instruction/ASID to pin the event to.
+ * TODO(am2419): Not sure if this is sensible, would be useful to have the event
+ * for consistency anyway, but what to do with multiple cpus?
+ */
+void qemu_log_instr_start(CPUArchState *env, uint32_t mode, target_ulong pc)
+{
+    cpu_log_instr_info_t *log = target_cpu_get_log(env);
+    uint32_t enabled = qemu_loglevel & INSTR_LOG_MASK;
+
+    log_assert(log != NULL && "Invalid log buffer");
+    log_assert((mode & CPU_LOG_INSTR) ^ (mode & CPU_LOG_CVTRACE) &&
+        "Instruction logging must be enabled in either INSTR or CVTRACE mode");
+
+    if ((mode & CPU_LOG_USER_ONLY)) {
+        // TODO(am2419): looks like translator_ops::tb_in_user_mode(db, cpu) does
```

Comment:
I added that and it will only work for RISCV+MIPS. Not sure if there is a better way of checking.